blog/2020-11-10-streaming-vitess-at-bolt.adoc

```diff
@@ -0,0 +1,181 @@
+* The change events include not only inserts and updates, but also deletes.
+* Probably one of the biggest advantages is that the change events contain the schema of each table. So you don’t have to worry about fetching each table’s schema in advance (by,  for example, parsing DDLs or querying the table’s definition).
+* The change events have VGTID included, which the CDC process can store and use as the offset from where to restart the CDC process next time.
+* Also importantly, VStream is designed to work well with Vitess operations such as link:https://vitess.io/docs/user-guides/resharding/[Resharding] and link:https://vitess.io/docs/user-guides/move-tables/[Moving Tables].
+
+There are also some disadvantages:
+
+* Although it includes table schemas, some important information is still missing. For example, the `Enum` and `Set` column types don’t provide all the allowed values yet. This should be fixed in the next major release (Vitess 9) though.
+* Since VStream is a gRPC service, we cannot use the Debezium MySQL Connector out-of-the-box. However, it is quite straightforward to implement the gRPC client in other languages.
+
+All things considered, we’ve decided to use VStream gRPC to capture change events from Vitess and implement our Vitess Connector based on all the best practices of Debezium.
+
+== Vitess Connector Deep Dive and Open Source
+
+After we’ve decided to implement our Vitess Connector, we started looking into the implementation details of various Debezium source connectors (MySQL, Postgres, SQLServer), to borrow some ideas. Almost all of them are implemented using a common Connector Development Framework. So it was clear we should develop the Vitess connector on top of it. Given we are very active users of MySql Connector and we benefit from it being open-sourced by having an ability to contribute to it things we were missing ourselves, we decided we want to give back to community and open-source Vitess source connector code-base under the Debezium umbrella. Please feel free to learn more at link:https://github.com/debezium/debezium-connector-vitess/[Debezium Connector Vitess]. We welcome and value any contributions.
+
+At a high level, as you can see below, connector instances are created in Kafka Connect workers. At the time of writing, you have two options to configure the connector to read from Vitess:
+
+*Option 1 (recommended):*
+
+As shown in _Figure 3_, each connector captures change events from all shards in a specific keyspace. If the keyspace is not sharded, the connector can still capture change events from the only shard in the keyspace. When it’s the first time that the connector starts, it reads from the current VGTID position of all shards in the keyspace. Because it subscribes to all shards, it continuously captures change events from all shards and sends them to Kafka. It automatically supports Vitess Reshard operation, there is no data loss, nor duplication.
```

Comment:
```suggestion
As shown in _Figure 3_, each connector captures change events from all shards in a specific keyspace. If the keyspace is not sharded, the connector can still capture change events from the only shard in the keyspace. When it’s the first time that the connector starts, it reads from the current VGTID position of all shards in the keyspace. Because it subscribes to all shards, it continuously captures change events from all shards and sends them to Kafka. It automatically supports the Vitess Reshard operation, there is no data loss, nor duplication.
```